File: app/src/main/java/com/organizer/layouts/MainLayout.java
package com.organizer.layouts;

import android.content.Context;
import android.graphics.Rect;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.AnimationUtils;
import android.view.animation.LinearInterpolator;
import android.view.animation.RotateAnimation;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.organizer.MainActivity;
import com.organizer.R;
import com.organizer.layouts.add.AddEventLayout;
import com.organizer.layouts.add.AddLayout;
import com.organizer.layouts.add.AddTaskLayout;
import com.organizer.layouts.calendar.CalendarLayout;
import com.organizer.layouts.todo.ToDoFolderLayout;
import com.organizer.layouts.todo.ToDoLayout;

public class MainLayout extends BaseLayout
{
    int scrollChildCount;
    private int startingChild = 1;
    private int currentChild;
    HorizontalScrollLayout toDoScrollLayout;
    HorizontalScrollLayout toDoFolderScrollLayout;
    ToDoLayout toDoLayout;
    ToDoFolderLayout toDoFolderLayout;
    CalendarLayout calendarLayout;
    FloatingActionButton addButton;
    AddLayout addLayout;
    private final Rect childRect = new Rect();
    GestureDetector gestureDetector = null;
    View.OnTouchListener touchListener = new View.OnTouchListener()
    {
        @Override
        public boolean onTouch(View view, MotionEvent motionEvent)
        {
            return gestureDetector.onTouchEvent(motionEvent) && view.isClickable();
        }
    };
    
    public MainLayout(Context context)
    {
        super(context);
        setGestureListener();
        
        toDoLayout = new ToDoLayout(context);
        
        toDoScrollLayout = new HorizontalScrollLayout(context);
        toDoScrollLayout.addContentView(toDoLayout);
        addView(toDoScrollLayout);
        
        calendarLayout = new CalendarLayout(context);
        addView(calendarLayout);
        
        scrollChildCount = getChildCount();
        for (int i = 0; i < scrollChildCount; i++)
        {
            getChildAt(i).setOnTouchListener(touchListener);
            getChildAt(i).setVisibility(View.INVISIBLE);
        }
        if (startingChild < 0 || startingChild >= scrollChildCount)
        {
            startingChild = 0;
        }
        currentChild = startingChild;
        getChildAt(startingChild).setVisibility(View.VISIBLE);
        
        toDoFolderLayout = new ToDoFolderLayout(context);
        
        toDoFolderScrollLayout = new HorizontalScrollLayout(context);
        toDoFolderScrollLayout.setVisibility(INVISIBLE);
        toDoFolderScrollLayout.addContentView(toDoFolderLayout);
        addView(toDoFolderScrollLayout);
        
        addButton = new FloatingActionButton(context);
        addButton.setImageResource(R.drawable.add_button);
        addButton.setOnClickListener(new OnClickListener()
        {
            @Override
            public void onClick(View view)
            {
                toggleAddLayout(addLayout.getVisibility() == INVISIBLE);
            }
        });
        addView(addButton);
        
        addLayout = new AddLayout(context);
        addLayout.setVisibility(INVISIBLE);
        addView(addLayout);
    }
    
    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom)
    {
        int width = right - left;
        int height = bottom - top;
        
        for (int i = 0; i < getChildCount(); i++)
        {
            getChildAt(i).layout(0, 0, width, height);
        }
        
        int margin = MainActivity.getDisplayMetricsController().dpToPx(16);
        childRect.left = width - addButton.getMeasuredWidth() - margin;
        childRect.top = height - addButton.getMeasuredHeight() - margin;
        childRect.right = width - margin;
        childRect.bottom = height - margin;
        addButton.layout(childRect.left, childRect.top, childRect.right, childRect.bottom);
    }
    
    public void toggleAddLayout(boolean visibility)
    {
        if (visibility)
        {
            toDoScrollLayout.setScrollable(false);
            toDoScrollLayout.setClickable(false);
            
            addLayout.setVisibility(VISIBLE);
            if (getChildAt(currentChild).equals(toDoScrollLayout))
            {
                addLayout.show(AddTaskLayout.class);
            }
            else if (getChildAt(currentChild).equals(calendarLayout))
            {
                addLayout.show(AddEventLayout.class);
            }
            
            RotateAnimation rotate = new RotateAnimation(0, 135, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
            rotate.setDuration(200);
            rotate.setFillAfter(true);
            rotate.setInterpolator(new LinearInterpolator());
            addButton.startAnimation(rotate);
        }
        else
        {
            toDoScrollLayout.setScrollable(true);
            toDoScrollLayout.setClickable(true);
            
            addLayout.setVisibility(INVISIBLE);
            
            RotateAnimation rotate = new RotateAnimation(135, 0, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
            rotate.setDuration(200);
            rotate.setFillAfter(true);
            rotate.setInterpolator(new LinearInterpolator());
            addButton.startAnimation(rotate);
        }
    }
    
    public HorizontalScrollLayout getToDoScrollLayout()
    {
        return toDoScrollLayout;
    }
    
    public HorizontalScrollLayout getToDoFolderScrollLayout()
    {
        return toDoFolderScrollLayout;
    }
    
    public ToDoLayout getToDoLayout()
    {
        return toDoLayout;
    }
    
    public ToDoFolderLayout getToDoFolderLayout()
    {
        return toDoFolderLayout;
    }
    
    public CalendarLayout getCalendarLayout()
    {
        return calendarLayout;
    }
    
    private void setGestureListener()
    {
        gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener()
        {
            private static final int SWIPE_THRESHOLD = 100;
            private static final int SWIPE_VELOCITY_THRESHOLD = 100;
            
            @Override
            public boolean onDown(MotionEvent e)
            {
                return true;
            }
            
            @Override
            public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY)
            {
                boolean result = false;
                try
                {
                    float diffY = e2.getY() - e1.getY();
                    float diffX = e2.getX() - e1.getX();
                    if (Math.abs(diffY) > SWIPE_THRESHOLD && Math.abs(velocityY) > SWIPE_VELOCITY_THRESHOLD && Math.abs(diffY) > Math.abs(diffX))
                    {
                        if (diffY > 0)
                        {
                            onSwipeDown();
                        }
                        else
                        {
                            onSwipeUp();
                        }
                        result = true;
                    }
                }
                catch (Exception exception)
                {
                    exception.printStackTrace();
                }
                return result;
            }
            
            void onSwipeUp()
            {
                if (currentChild < scrollChildCount - 1)
                {
                    AnimationSet animation = new AnimationSet(false);
                    final View child1 = getChildAt(currentChild);
                    final View child2 = getChildAt(currentChild + 1);
                    
                    animation.addAnimation(AnimationUtils.loadAnimation(context, R.anim.fadeoutup));
                    animation.setAnimationListener(new Animation.AnimationListener()
                    {
                        @Override
                        public void onAnimationStart(Animation animation)
                        {
                            child1.setVisibility(View.INVISIBLE);
                        }
                        
                        @Override
                        public void onAnimationEnd(Animation animation)
                        {
                            child2.setVisibility(View.VISIBLE);
                        }
                        
                        @Override
                        public void onAnimationRepeat(Animation animation)
                        {
                        }
                    });
                    child1.startAnimation(animation);
                    
                    currentChild++;
                }
            }
            
            void onSwipeDown()
            {
                if (currentChild > 0)
                {
                    AnimationSet animation = new AnimationSet(false);
                    final View child1 = getChildAt(currentChild);
                    final View child2 = getChildAt(currentChild - 1);
                    
                    animation.addAnimation(AnimationUtils.loadAnimation(context, R.anim.fadeoutdown));
                    animation.setAnimationListener(new Animation.AnimationListener()
                    {
                        @Override
                        public void onAnimationStart(Animation animation)
                        {
                            child1.setVisibility(View.INVISIBLE);
                        }
                        
                        @Override
                        public void onAnimationEnd(Animation animation)
                        {
                            child2.setVisibility(View.VISIBLE);
                        }
                        
                        @Override
                        public void onAnimationRepeat(Animation animation)
                        {
                        }
                    });
                    child1.startAnimation(animation);
                    
                    currentChild--;
                }
            }
        });
    }
}

File: app/src/main/java/com/organizer/layouts/MainLayout.kt
package com.organizer.layouts

import android.annotation.SuppressLint
import android.graphics.Rect
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View.OnTouchListener
import android.view.animation.*
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.organizer.*
import com.organizer.MainActivity.Companion.inject
import com.organizer.layouts.add.*
import com.organizer.layouts.calendar.CalendarLayout
import com.organizer.layouts.todo.ToDoFolderLayout
import com.organizer.layouts.todo.ToDoLayout
import kotlin.math.abs

@SuppressLint("ViewConstructor")
class MainLayout : BaseLayout() {
    private val displayMetricsController: DisplayMetricsController by inject()
    var scrollChildCount: Int
    private var startingChild = 1
    private var currentChild: Int
    var toDoScrollLayout: HorizontalScrollLayout
    var toDoFolderScrollLayout: HorizontalScrollLayout
    var toDoLayout: ToDoLayout
    var toDoFolderLayout: ToDoFolderLayout
    var calendarLayout: CalendarLayout
    var addButton: FloatingActionButton
    private var addLayout: AddLayout
    private val childRect = Rect()
    var gestureDetector: GestureDetector? = null
    private var touchListener = OnTouchListener { view, motionEvent ->
        super.performClick()
        gestureDetector!!.onTouchEvent(motionEvent) && view.isClickable
    }

    init {
        setGestureListener()

        toDoLayout = ToDoLayout()
        toDoScrollLayout = HorizontalScrollLayout()
        toDoScrollLayout.addContentView(toDoLayout)
        addView(toDoScrollLayout)

        calendarLayout = CalendarLayout()
        addView(calendarLayout)

        scrollChildCount = childCount
        for (i in 0 until scrollChildCount) {
            getChildAt(i).setOnTouchListener(touchListener)
            getChildAt(i).visibility = INVISIBLE
        }
        if (startingChild < 0 || startingChild >= scrollChildCount) {
            startingChild = 0
        }
        currentChild = startingChild
        getChildAt(startingChild).visibility = VISIBLE
        toDoFolderLayout = ToDoFolderLayout()
        toDoFolderScrollLayout = HorizontalScrollLayout()
        toDoFolderScrollLayout.visibility = INVISIBLE
        toDoFolderScrollLayout.addContentView(toDoFolderLayout)
        addView(toDoFolderScrollLayout)

        addLayout = AddLayout()
        addLayout.visibility = INVISIBLE
        addView(addLayout)

        addButton = FloatingActionButton(context)
        addButton.setImageResource(R.drawable.add_button)
        addButton.setOnClickListener { toggleAddLayout(addLayout.visibility == INVISIBLE) }
        addView(addButton)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val width = right - left
        val height = bottom - top
        for (i in 0 until childCount) {
            getChildAt(i).layout(0, 0, width, height)
        }
        val margin = displayMetricsController.dpToPx(16f)
        childRect.left = width - addButton.measuredWidth - margin
        childRect.top = height - addButton.measuredHeight - margin
        childRect.right = width - margin
        childRect.bottom = height - margin
        addButton.layout(childRect.left, childRect.top, childRect.right, childRect.bottom)
    }

    fun toggleAddLayout(visibility: Boolean) {
        if (visibility) {
            toDoScrollLayout.isScrollable = false
            toDoScrollLayout.isClickable = false
            addLayout.visibility = VISIBLE
            if (getChildAt(currentChild) == toDoScrollLayout) {
                addLayout.show(AddTaskLayout::class.java)
            } else if (getChildAt(currentChild) == calendarLayout) {
                addLayout.show(AddEventLayout::class.java)
            }
            val rotate = RotateAnimation(
                0F,
                135F,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f
            )
            rotate.duration = 200
            rotate.fillAfter = true
            rotate.interpolator = LinearInterpolator()
            addButton.startAnimation(rotate)
        } else {
            toDoScrollLayout.isScrollable = true
            toDoScrollLayout.isClickable = true
            addLayout.visibility = INVISIBLE
            val rotate = RotateAnimation(
                135F,
                0F,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f
            )
            rotate.duration = 200
            rotate.fillAfter = true
            rotate.interpolator = LinearInterpolator()
            addButton.startAnimation(rotate)
        }
    }

    private fun setGestureListener() {
        gestureDetector = GestureDetector(context, object : SimpleOnGestureListener() {
            val SWIPE_THRESHOLD = 100
            val SWIPE_VELOCITY_THRESHOLD = 100
            override fun onDown(e: MotionEvent): Boolean {
                return true
            }

            override fun onFling(
                e1: MotionEvent,
                e2: MotionEvent,
                velocityX: Float,
                velocityY: Float
            ): Boolean {
                var result = false
                try {
                    val diffY = e2.y - e1.y
                    val diffX = e2.x - e1.x
                    if (abs(diffY) > SWIPE_THRESHOLD && abs(velocityY) > SWIPE_VELOCITY_THRESHOLD && abs(
                            diffY
                        ) > abs(diffX)
                    ) {
                        if (diffY > 0) {
                            onSwipeDown()
                        } else {
                            onSwipeUp()
                        }
                        result = true
                    }
                } catch (exception: Exception) {
                    exception.printStackTrace()
                }
                return result
            }

            fun onSwipeUp() {
                if (currentChild < scrollChildCount - 1) {
                    val animation = AnimationSet(false)
                    val child1 = getChildAt(currentChild)
                    val child2 = getChildAt(currentChild + 1)
                    animation.addAnimation(AnimationUtils.loadAnimation(context, R.anim.fadeoutup))
                    animation.setAnimationListener(object : Animation.AnimationListener {
                        override fun onAnimationStart(animation: Animation) {
                            child1.visibility = INVISIBLE
                        }

                        override fun onAnimationEnd(animation: Animation) {
                            child2.visibility = VISIBLE
                        }

                        override fun onAnimationRepeat(animation: Animation) {}
                    })
                    child1.startAnimation(animation)
                    currentChild++
                }
            }

            fun onSwipeDown() {
                if (currentChild > 0) {
                    val animation = AnimationSet(false)
                    val child1 = getChildAt(currentChild)
                    val child2 = getChildAt(currentChild - 1)
                    animation.addAnimation(
                        AnimationUtils.loadAnimation(
                            context,
                            R.anim.fadeoutdown
                        )
                    )
                    animation.setAnimationListener(object : Animation.AnimationListener {
                        override fun onAnimationStart(animation: Animation) {
                            child1.visibility = INVISIBLE
                        }

                        override fun onAnimationEnd(animation: Animation) {
                            child2.visibility = VISIBLE
                        }

                        override fun onAnimationRepeat(animation: Animation) {}
                    })
                    child1.startAnimation(animation)
                    currentChild--
                }
            }
        })
    }
}
